core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/GoogleServicesDAO.scala

```diff
@@ -123,6 +123,10 @@ abstract class GoogleServicesDAO(groupsPrefix: String) extends ErrorReportable {
     */
   def listBillingAccountsUsingServiceCredential(implicit executionContext: ExecutionContext): Future[Seq[RawlsBillingAccount]]
 
+  def getProjectBillingAccount(projectName: RawlsBillingProjectName, userInfo: UserInfo)(implicit executionContext: ExecutionContext): Future[Option[String]]
```

Comment:
We are trying to be more explicit whenever we use the overloaded term, `project`, and say either `billingProject` or `googleProject`.  So this signature would be something like:

```suggestion
  def getBillingProjectsBillingAccountName(billingProjectName: RawlsBillingProjectName, userInfo: UserInfo)(implicit executionContext: ExecutionContext): Future[Option[String]]
```

Or maybe:
```suggestion
  def getBillingAccountNameForBillingProject(billingProjectName: RawlsBillingProjectName, userInfo: UserInfo)(implicit executionContext: ExecutionContext): Future[Option[String]]
```

In a more general sense, I think we should be very precise with our naming here because I got confused reading this at first about what it was returning.  The fact that this is just returning a Billing Account _Name_ threw me off since we are usually dealing with IDs.

Reply:
And I haven't gotten into the calling code yet, but this method is going to need to change or it is going to break when we get into PPW.  The reason being that in PPW, each Workspace will have its own Google Project.  All of the Workspaces (and therefore, Google Projects) _should_ have the same Billing Account set on them as specified by the Billing Project.  However, it is not guaranteed that this will be the case.  Updating the Billing Account for all of the Workspaces/GoogleProjects in a Billing Project is going to become an asynchronous operation. 

There is a possibility that at least one Google Projects errors out when updating the Billing Account.  Furthermore, during an update operation, there will naturally be some Google Projects with the old Billing Account still set and some that have already been updated with the new Billing Account.  

TL;DR - After PPW launches, it won't make sense to ask `GoogleServicesDAO` to get the Billing Account for a Billing Project, because it is possible that a Billing Project may have multiple different Billing Accounts specified on the Google Projects within it.  That is not a state we want the data to remain in, but it is nonetheless a valid state.

So although weird, maybe we add this method and mark it as deprecated right away or something.  Something to tell us that this thing is going to need to change real soon.  And maybe a Jira ticket is needed to update this in the PPW branch.

Reply:
I've updated the code a little bit to try to clarify this stuff more and referenced a new Jira ticket that I've created.

As discussed over Zoom, it definitely feels like there's going to need to be some refactoring once PPW goes live. There are some things that can be prepped ahead of time but I think there are some re-factorings that are going to be inevitable. 

The thing that concerns me the most is this multi-billing-account state that you've described. If it's the case that the Google projects within a billing project can temporarily map to different billing accounts, then I'm not sure how this is going to work without significant refactoring.

This will require product input but IMO a reasonable way to deal with this inconsistent state would be to prevent users from setting a spend configuration on a billing project that has inconsistent billing accounts across its Google projects. It's already going to be the case that the spend configuration will be cleared when the billing account changes, so there's already some "downtime" on spend reporting anyway in these circumstances.

I offer this solution knowing that I don't have much insight into how frequent a billing project could end up in an inconsistent state and for how long it could last.